src/python/pants/engine/internals/selectors.py

```diff
@@ -233,183 +227,172 @@ def __await__(self) -> Generator[Tuple[Get, ...], None, Tuple]:
+    __get3: Get[_Out3, _In3],
+    __get4: Get[_Out4, _In4],
+) -> Tuple[_Out0, _Out1, _Out2, _Out3, _Out4]:
     ...
 
 
 @overload
 async def MultiGet(  # noqa: F811
-    __get0: Get[_P0, _S0],
-    __get1: Get[_P1, _S1],
-    __get2: Get[_P2, _S2],
-    __get3: Get[_P3, _S3],
-) -> Tuple[_P0, _P1, _P2, _P3]:
+    __get0: Get[_Out0, _In0],
+    __get1: Get[_Out1, _In1],
+    __get2: Get[_Out2, _In2],
+    __get3: Get[_Out3, _In3],
+) -> Tuple[_Out0, _Out1, _Out2, _Out3]:
     ...
 
 
 @overload
 async def MultiGet(  # noqa: F811
-    __get0: Get[_P0, _S0], __get1: Get[_P1, _S1], __get2: Get[_P2, _S2]
-) -> Tuple[_P0, _P1, _P2]:
+    __get0: Get[_Out0, _In0], __get1: Get[_Out1, _In1], __get2: Get[_Out2, _In2]
+) -> Tuple[_Out0, _Out1, _Out2]:
     ...
 
 
 @overload
-async def MultiGet(__get0: Get[_P0, _S0], __get1: Get[_P1, _S1]) -> Tuple[_P0, _P1]:  # noqa: F811
+async def MultiGet(
+    __get0: Get[_Out0, _In0], __get1: Get[_Out1, _In1]
+) -> Tuple[_Out0, _Out1]:  # noqa: F811
     ...
 
 
 async def MultiGet(  # noqa: F811
-    __arg0: Union[Iterable[Get[_ProductType, _SubjectType]], Get[_P0, _S0]],
-    __arg1: Optional[Get[_P1, _S1]] = None,
-    __arg2: Optional[Get[_P2, _S2]] = None,
-    __arg3: Optional[Get[_P3, _S3]] = None,
-    __arg4: Optional[Get[_P4, _S4]] = None,
-    __arg5: Optional[Get[_P5, _S5]] = None,
-    __arg6: Optional[Get[_P6, _S6]] = None,
-    __arg7: Optional[Get[_P7, _S7]] = None,
-    __arg8: Optional[Get[_P8, _S8]] = None,
-    __arg9: Optional[Get[_P9, _S9]] = None,
+    __arg0: Union[Iterable[Get[_Output, _Input]], Get[_Out0, _In0]],
+    __arg1: Optional[Get[_Out1, _In1]] = None,
+    __arg2: Optional[Get[_Out2, _In2]] = None,
+    __arg3: Optional[Get[_Out3, _In3]] = None,
+    __arg4: Optional[Get[_Out4, _In4]] = None,
+    __arg5: Optional[Get[_Out5, _In5]] = None,
+    __arg6: Optional[Get[_Out6, _In6]] = None,
+    __arg7: Optional[Get[_Out7, _In7]] = None,
+    __arg8: Optional[Get[_Out8, _In8]] = None,
+    __arg9: Optional[Get[_Out9, _In9]] = None,
 ) -> Union[
-    Tuple[_ProductType, ...],
-    Tuple[_P0, _P1, _P2, _P3, _P4, _P5, _P6, _P7, _P8, _P9],
-    Tuple[_P0, _P1, _P2, _P3, _P4, _P5, _P6, _P7, _P8],
-    Tuple[_P0, _P1, _P2, _P3, _P4, _P5, _P6, _P7],
-    Tuple[_P0, _P1, _P2, _P3, _P4, _P5, _P6],
-    Tuple[_P0, _P1, _P2, _P3, _P4, _P5],
-    Tuple[_P0, _P1, _P2, _P3, _P4],
-    Tuple[_P0, _P1, _P2, _P3],
-    Tuple[_P0, _P1, _P2],
-    Tuple[_P0, _P1],
-    Tuple[_P0],
+    Tuple[_Output, ...],
+    Tuple[_Out0, _Out1, _Out2, _Out3, _Out4, _Out5, _Out6, _Out7, _Out8, _Out9],
+    Tuple[_Out0, _Out1, _Out2, _Out3, _Out4, _Out5, _Out6, _Out7, _Out8],
+    Tuple[_Out0, _Out1, _Out2, _Out3, _Out4, _Out5, _Out6, _Out7],
+    Tuple[_Out0, _Out1, _Out2, _Out3, _Out4, _Out5, _Out6],
+    Tuple[_Out0, _Out1, _Out2, _Out3, _Out4, _Out5],
+    Tuple[_Out0, _Out1, _Out2, _Out3, _Out4],
+    Tuple[_Out0, _Out1, _Out2, _Out3],
+    Tuple[_Out0, _Out1, _Out2],
+    Tuple[_Out0, _Out1],
+    Tuple[_Out0],
```

Comment:
Now THIS looks like something I've seen in the scala compiler, lol. The `@overload` mechanism used here is really cool.